### PR TITLE
Multiple constitutionality proposal display fix  (#4111)

### PR DESCRIPTION
**Pull request by chrlschwb**
fixes issue #4111

### CODE REVIEW

File: packages/ui/src/proposals/hooks/useProposal.ts
Comment (chrlschwb) [inline packages/ui/src/proposals/hooks/useProposal.ts:15]: This query data doesn't contain all proposal data. In case of 3 constitutionals, this only displays 2 constitutionals.
This is evidently an error that is created during proposal creation.

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (gyroflaw) [inline packages/ui/src/proposals/components/ProposalStages.tsx:40]: Nice,

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (gyroflaw) [inline packages/ui/src/proposals/components/ProposalStages.tsx:60]: LGTM

**Review by gyroflaw — APPROVED** [review]
LGTM

**Review by traumschule — APPROVED** [review]
LGTM! As next step we can add expected election and anticipated execution dates via separate PR, for example in `HISTORY` section. cc @bedeho 

const-tooltip

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (thesan) [inline packages/ui/src/proposals/components/ProposalStages.tsx:49]: Here the "disabled" tabs are those with `typeof onClick === 'undefined'`... Which is not very readable code.

IMO the best would be to change the type of `rounds`:
from: `{ icon: boolean, onClick: undefined | () => void }` to: `'approved' | 'rejected' | 'deciding' | 'disabled'`.
So this would become something like:
```tsx
{rounds.map((status, round) => {
  const isDisabled = status === 'disabled'
  const isActive = round === value
  const onClick = isDisabled ? undefined : () => onChange(round)
  const icon = iconMap[status]

  return (
  <Tooltip key={round} tooltipText={isDisabled ? "text1..." : "text2..."}>
    <TabContainer active={isActive} disabled={isDisabled} onClick={onClick}>
      {icon}
      <spann>
        ...
```

With `iconMap` defined on the module scope as:
```ts
const iconMap = { approved: <CheckboxIcon />, rejected: <CrossIcon />, deciding: undefined, disabled: undefined }
```
or even as a function with a switch...

**Review by thesan — CHANGES_REQUESTED** [review]
Thank you for the PR @chrlschwb. It still need some work on the behavior but I think some code improvement on the previous solution would be helpful too.

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (chrlschwb) [reply packages/ui/src/proposals/components/ProposalStages.tsx:40]: typeof constitutionality === "string" means if constitutionality === "-" then undefined else if constitutionality === number then constituonality.

Comment (thesan) [reply packages/ui/src/proposals/components/ProposalStages.tsx]: I see. That was a bad solution on my end :sweat: . 
It should be:
```ts
export interface ProposalStagesProps extends ControlProps<number> {
  rounds: VotingRound[]
  status: ProposalStatus
  constitutionality?: number
}

export const ProposalStages = ({ rounds, status, constitutionality, value, onChange }: ProposalStagesProps) => {
```

and `Council approvals {round + 1}/{constitutionality ?? '-'}`

This way the return becomes:
```ts
repeat(
  () => ...,
  constitutionality ?? 1
)
```
no need for `totalRoundCount`

Comment (chrlschwb) [reply packages/ui/src/proposals/components/ProposalStages.tsx]: image
There are errors when I try to change the code as you suggested.

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (traumschule) [reply packages/ui/src/proposals/components/ProposalStages.tsx:50]: The [type of `status` is `ProposalStatus`](https://github.com/Joystream/pioneer/blob/6851ba0f445c676586eea8260017142ddc317e5c/packages/ui/src/proposals/components/ProposalStages.tsx#L11) defined [here](https://github.com/Joystream/pioneer/blob/dev/packages/ui/src/proposals/types/proposals.ts#L14) as `String`.

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (thesan) [reply packages/ui/src/proposals/components/ProposalStages.tsx]: Yes it looks like I wrote `rounds` instead of `updates`. But still the current code should be changed since it doesn't achieve what we want on the deciding stage. Because now we want to always show all the tabs (although some of them might be disabled). The number of tabs will always be `constitutionality`. So there should be only one `repeat`.

About my second comment I was proposing to change the type of the items returned by this repeat from: `{ icon: boolean, onClick: undefined | () => void }` **to**: `'approved' | 'rejected' | 'deciding' | 'disabled'` in order to make the code more readable.

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (thesan) [reply packages/ui/src/proposals/components/ProposalStages.tsx:50]: @chrlschwb You're right about it currently being an object but this why I suggested to change it to: `'approved' | 'rejected' | 'deciding' | 'disabled'`. I guess calling the variable `status` is confusing maybe `roundStatus` would be better.

File: packages/ui/src/council/components/election/CandidateVote/CandidateVote.tsx
Comment (traumschule) [inline packages/ui/src/council/components/election/CandidateVote/CandidateVote.tsx:63]: This section is superseded by #4155 and could be merged here.

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (traumschule) [inline packages/ui/src/proposals/components/ProposalStages.tsx:45]: Referring to the [original comment](https://github.com/Joystream/pioneer/pull/4119#discussion_r1091047618): `dormant` is the phase after approval until the next council. However if a tab `isDisabled` depends on content of `onClick`:
> Here the "disabled" tabs are in fact those with typeof onClick === 'undefined' (not round !== value), which is not very readable code.

**Review by traumschule — CHANGES_REQUESTED** [review]
Please correct / explain.

**Review by thesan — CHANGES_REQUESTED** [review]

**Comment by traumschule** [comment]
Clicking on the tab does not have an effect. As i understand it the votes by that council should be shown.

**Comment by traumschule** [comment]
This looks very good already. History shows all stages with date and it's possible to click each tab to review votes and rationales.

It is just unclear why after 3x3 CM approved the proposal is still deciding. Like what is the next step? The answer isn't obvious in either stage. Suggest to merge soon and create follow-ups.

consti

@bedeho wdyt? https://dao-git-fork-chrlschwb-fix-4111-proposalsdisplay-joystream.vercel.app/#/proposals/preview/63

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (traumschule) [inline packages/ui/src/proposals/components/ProposalStages.tsx:16]: Maybe these should be included from `types.ts`? I checked the code and there are [multiple locations](https://github.com/search?q=org%3AJoystream+dormant&type=code) that hard-code possible stages. @thesan wdyt? For example [here](https://github.com/Joystream/pioneer/blob/bb6f63bd6455c812b8934d96f0b964bb4192d782/packages/ui/src/proposals/types/proposals.ts#L27):
> type CurrentProposalStatus = Extract<ProposalStatus, 'deciding' | 'gracing' | 'dormant'>

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (traumschule) [inline packages/ui/src/proposals/components/ProposalStages.tsx:59]: minor nits:
- check for `isDisabled` (not inverse)
- define strings / test earlier to make component easier to read:
>  <Tooltip key={round} tooltipText={isDisabled ? tooltip.disabled : tooltip.active}>

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (traumschule) [inline packages/ui/src/proposals/components/ProposalStages.tsx:61]:
```suggestion
                : 'The number of councils that must approve the proposal in a row before it has its intended effect is more than one. The execution is triggered after the last council completes the voting. Votes for each council are displayed in a separate tab.'
```
Question: In above screenshot voting is complete but status is "Deciding". What am i missing?
> The execution is triggered after the last council completes the voting.

**Review by traumschule — APPROVED** [review]
LGTM, some room to refactor though.

**Comment by bedeho** [comment]
I am a bit confused, why do I see voting like this for `2/3`

Screenshot 2023-02-19 at 22 16 51


is this a bug? seems to combine voting for differnt councils which makes 0 sense, we should show voting pattern for the council selected only.

**Comment by bedeho** [comment]
Also, what is going on here with being in deciding ?

Screenshot 2023-02-19 at 22 17 54

**Comment by gyroflaw** [comment]
> I am a bit confused, why do I see voting like this for `2/3`
> 
> Screenshot 2023-02-19 at 22 16 51
> 
> is this a bug? seems to combine voting for differnt councils which makes 0 sense, we should show voting pattern for the council selected only.

I think this should be handled in another PR :)

File: packages/ui/src/app/pages/Proposals/ProposalPreview.tsx
Comment (thesan) [inline packages/ui/src/app/pages/Proposals/ProposalPreview.tsx:165]: I don't think this should be changed. This is the status of the proposal so "status" is correct

File: packages/ui/src/council/components/election/CandidateVote/CandidateVote.tsx
Comment (thesan) [inline packages/ui/src/council/components/election/CandidateVote/CandidateVote.tsx:93]: This look like a mistake

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (thesan) [inline packages/ui/src/proposals/components/ProposalStages.tsx:11]: Again this should remain "status"

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (thesan) [reply packages/ui/src/proposals/components/ProposalStages.tsx:16]: This type is just for this file. It's the status of a round which is different from the status of the proposal this is why I think it should stay here and not be exported.

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (thesan) [inline packages/ui/src/proposals/components/ProposalStages.tsx:24]:
```suggestion
    const onGoing = lastUpdate === status
```

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (thesan) [inline packages/ui/src/proposals/components/ProposalStages.tsx:20]: :+1:

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (thesan) [reply packages/ui/src/proposals/components/ProposalStages.tsx:61]: @traumschule it shows 3 votes but they are the votes from the first round so it looks to me like it the issue #4201

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (thesan) [inline packages/ui/src/proposals/components/ProposalStages.tsx:32]: It should be `decidingCount - 1`:
```suggestion
      if (round < decidingCount - 1) {
        return 'approved'
      } else if (round > decidingCount - 1) {
```

**Review by thesan — CHANGES_REQUESTED** [review]

File: packages/ui/src/proposals/components/ProposalStages.tsx
Comment (chrlschwb) [reply packages/ui/src/proposals/components/ProposalStages.tsx]: Ok. I will do.

**Comment by traumschule** [comment]
```
 | Summary of all failing tests
 | FAIL test/council/pages/PastCouncil.test.tsx (13.677 s)
 |   ● UI: Past Council page › Council found › Renders
 | 
 |     Timed out in waitForElementToBeRemoved.
```

**Review by thesan — APPROVED** [review]
@chrlschwb I implemented some changes since this PR went through a lot of review cycle already. I also included @traumschule requests. @traumschule these were good suggestion, please don't approve PRs when you found some code quality and typos issues.

I realized there was a confusing point about the `updates` not including the initial deciding round which may have contributed to the confusion on this PR.
It also results in this minor bug #4213.

**Pull request merged**
